src/Features/VolumetricLighting.cpp

```diff
@@ -66,15 +159,83 @@ void VolumetricLighting::DataLoaded()
+	gVolumetricLightingSizeLow = reinterpret_cast<TextureSize*>(REL::VariantOffset(0x3233090, 0x338C550, 0x3485630).address());
+	gVolumetricLightingSizeMedium = reinterpret_cast<TextureSize*>(REL::VariantOffset(0x323309C, 0x338C55C, 0x348563C).address());
+	gVolumetricLightingSizeHigh = reinterpret_cast<TextureSize*>(REL::VariantOffset(0x32330A8, 0x338C568, 0x3485648).address());
+	defaultSizeHigh = *gVolumetricLightingSizeHigh;
+}
+
+void VolumetricLighting::EarlyPrepass()
+{
+	const auto interiorCell = RE::TES::GetSingleton()->interiorCell;
+	const bool currentlyInInterior = interiorCell != nullptr;
+
+	if (initialised && currentlyInInterior == inInterior)
+		return;
+
+	initialised = true;
+	inInterior = currentlyInInterior;
+	inInteriorWithSunShadows = interiorCell && interiorCell->cellFlags.all(RE::TESObjectCELL::Flag::kIsInteriorCell, RE::TESObjectCELL::Flag::kShowSky, RE::TESObjectCELL::Flag::kUseSkyLighting);
+	SetupVL();
+}
+
+void VolumetricLighting::SetupVL()
+{
+	if (inInterior) {
+		if (globals::game::isVR)
+			SetBooleanSettings(hiddenVRSettings, GetName(), settings.InteriorEnabled && inInteriorWithSunShadows);
+		else
+			*bEnableVolumetricLighting = settings.InteriorEnabled && inInteriorWithSunShadows;
+		*gVolumetricLightingSizeHigh = static_cast<Quality>(settings.InteriorQuality) == Quality::Custom ? settings.InteriorCustomSize : defaultSizeHigh;
+		SetVLQuality(GetVLDescriptor(), settings.InteriorQuality);
+	} else {
+		if (globals::game::isVR)
+			SetBooleanSettings(hiddenVRSettings, GetName(), settings.ExteriorEnabled);
+		else
+			*bEnableVolumetricLighting = settings.ExteriorEnabled;
+		*gVolumetricLightingSizeHigh = static_cast<Quality>(settings.ExteriorQuality) == Quality::Custom ? settings.ExteriorCustomSize : defaultSizeHigh;
+		SetVLQuality(GetVLDescriptor(), settings.ExteriorQuality);
+	}
+}
+
+VolumetricLighting::VolumetricLightingDescriptor& VolumetricLighting::GetVLDescriptor()
+{
+	using func_t = decltype(&VolumetricLighting::GetVLDescriptor);
+	static REL::Relocation<func_t> func{ REL::RelocationID(100297, 107014) };
+	return func();
+}
+
+void VolumetricLighting::SetVLQuality(VolumetricLightingDescriptor& descriptor, const uint32_t quality)
+{
+	using func_t = decltype(&VolumetricLighting::SetVLQuality);
+	static REL::Relocation<func_t> func{ REL::RelocationID(100299, 107016).address() };
+	func(descriptor, std::clamp<uint32_t>(quality, 0, 2));
+}
+
+void VolumetricLighting::RenderVolumetricLighting(VolumetricLightingDescriptor* descriptor, RE::NiCamera* camera, bool flag)
+{
+	using func_t = decltype(&VolumetricLighting::RenderVolumetricLighting);
+	static REL::Relocation<func_t> func{ REL::RelocationID(100306, 0) };
+	func(descriptor, camera, flag);
 }
 
-void VolumetricLighting::Reset()
+void VolumetricLighting::RenderDepth::thunk()
 {
+	func();
+	if (*GetSingleton()->bEnableVolumetricLighting)
+		RenderVolumetricLighting(&GetVLDescriptor(), RE::Main::WorldRootCamera(), false);
```

Comment:
WorldRootCamera, would that worked as Global or it changing too often?

Reply:
I think I'd prefer to read it each frame here as it's performing essentially the same action as the original code and I'm not sure of the consequences when switching between first/third person or the map menu.